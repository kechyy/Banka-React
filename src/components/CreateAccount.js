import React, { Component } from 'react';
import { connect } from 'react-redux'; 
import {withRouter} from 'react-router-dom';
import AccountList from './AccountList';
import { getAccountAction, accountAction } from '../store/modules/bankAccount/actions';
import { profileAction } from '../store/modules/userProfile/actions';
import SideBar from './SideBar';
import MainFooter from './MainFooter';
import MainHeader from './MainHeader';

class CreateAccount extends Component  {
  constructor() {
    super();
    this.state = {
      title: 'Bank Account',
      type: '',
      errors: {
        type: '',
      }
    } 
    this.createNewAccount = this.createNewAccount.bind(this)
    this.handleInputFieldChange = this.handleInputFieldChange.bind(this )
  }

  createNewAccount(e) {
    e.preventDefault();
    const { type } = this.state;
    const { accountAction } = this.props;
    const errors = {};
    if (!type) errors.type = 'Account type cannot be empty';
    if (Object.keys(errors).length) {
      this.setState(prevState => ({
        ...prevState,
        errors: {
          ...errors
        }
      }));
      return;
    }
    accountAction({ type });
  }
  componentDidMount () {
    document.body.className = 'main-body';
    this.props.getAccountAction();
    this.props.profileAction(); 
  }

  handleInputFieldChange(event) {
    const {
      target: { name, value }
    } = event;
    this.setState(prevState => ({
      [name]: value,
      errors: {
        ...prevState.errors,
        [name]: ''
      }
    }));
  }
 
  

  render() {
    
     const {userProfile} = this.props;
    return (
      <div>
        <MainHeader />
        <main>
          <div className="wrapper ">
            <SideBar />
            <div className="right">
              <div className="row">
                <div className="col-12-lg mainContent">
                  <h2 id="contentMainTitle" >User Account Page</h2>
                  <header className="mainTitle">
                    <div className="row ">
                      <div className="col-12-md">
                        <h3 id="contentSubTitle" >{this.state.title}</h3>
                      </div>
                    </div>
                  </header>
                  <aside className="rightSideBar">
                    <div className="row" id="userMain">
                    <div style={{'width':'100%'}} >
                      <div className="col-8-md acctDetails">
                      <form onSubmit={this.createNewAccount}>
                        <div className="row createAcct">
                          <div className="col-6-md col-12-xs">
                            <label htmlFor="true">First Name</label>
                            <input type="text" className="input-block" defaultValue={userProfile.firstname} disabled />
                          </div>
                          <div className="col-6-md col-12-xs">
                            <label htmlFor="true">Last Name</label>
                            <input type="text" className="input-block" defaultValue={userProfile.lastname} disabled />
                          </div>
                          <div className="col-6-md col-12-xs">
                            <label htmlFor="true">Email Address</label>
                            <input type="email" className="input-block" defaultValue={userProfile.email} disabled />
                          </div>
                          <div className="col-6-md col-12-xs">
                            <label htmlFor="accountType" className="required">Account type <span>*</span></label>
                            <select className="input-block" required id="type" name="type" onChange={this.handleInputFieldChange}>
                              <option value="">--Select Account Type--</option>
                              <option value="savings">Savings</option>
                              <option value="current">Current</option>
                            </select>
                          </div>
                          <div className="col-6-md col-12-xs">
                            <input type="submit" className="btn btn-blue createAccountBtn" id="createAccountBtn" defaultValue="Create Account" />
                          </div>
                        </div>
                      </form>
                      <div className="row ">
                          <div className="col-12-xs col-12-md table">
                            <table cellPadding={0} cellSpacing={0}>
                              <thead>
                                <tr>
                                  <th>S/N</th>
                                  <th>Account Number</th>
                                  <th>Account type</th>
                                  <th>Account Balance</th>
                                  <th>Account Status</th>
                                  <th>Created Date</th>
                                </tr>
                              </thead>
                              <tbody>
                                { this.props.account.map((acct, y) => {
                                    return <AccountList accounts={acct} i={y}  />
                                })}
                              </tbody>
                            </table>
                          </div>
                          </div>
                        </div>
                      </div>
                    </div>
                  </aside>
                </div>
              </div> 
            </div>
          </div>
        </main>
       <MainFooter />
      </div>
    )
  }
}
const mapStateToProps = state => ({
  account: state.accountReducer.account,
  userProfile: state.profileReducer.profile
});
const mapDispatchToProps = {
  getAccountAction,  profileAction, accountAction
}
export default withRouter(connect(mapStateToProps, mapDispatchToProps)(CreateAccount));