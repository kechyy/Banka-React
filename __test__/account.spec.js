import React from 'react';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { mount, shallow } from './enzyme';
import CreateAccount from '../src/components/CreateAccount';
import { BrowserRouter as Router, Route } from 'react-router-dom';
import toJson from 'enzyme-to-json';
import { BrowserRouter, Switch } from 'react-router-dom';
import { Provider } from 'react-redux';


const mockStore = configureMockStore([thunk]);
let store;
const initialSate = {
  data: [],
  error: null,
  status: 'rest'
};

const props = {
  isAuthenticated: true,
  getAccountAction: jest.fn(),
  accountAction: jest.fn(),
  userProfile: jest.fn()
};

const renderWithEnzymes = state => {
  store = mockStore({
    authReducer: {
      error: null,
      isAuthenticated: false,
      user: {},
      status: 'rest'
    },
   
    accountReducer: {
      account: [],
      error: null,
      status: 'rest'
    },
    profileReducer: {
      profile: {},
      status: 'rest'
    },
    CreateAccount: {
      ...state
    }
  });
  return mount(
    <Provider store={store}>
      <Router>
        <CreateAccount {...props} />
      </Router>
    </Provider>
  );
};


describe('CreateAccount', () => {
  afterEach(() => {
    jest.clearAllMocks();
    store = null;
  });

  it('should have 4 input on create account', () => {
    const wrapper = renderWithEnzymes();

    expect(wrapper.find('input').length).toBe(4);
    expect(
      wrapper
        .find('input')
        .at(0)
        .props().type
    ).toBe('text');
    expect(
      wrapper
        .find('input')
        .at(1)
        .props().type
    ).toBe('text');
    expect(
      wrapper
        .find('input')
        .at(2)
        .props().type
    ).toBe('email');
    expect(
      wrapper
        .find('input')
        .at(3)
        .props().type
    ).toBe('submit');
  });

  it('should have 1 select', () => {
    const wrapper = renderWithEnzymes();
    const field = wrapper.find('select').first();
    expect(field.exists()).toBe(true);
    expect(field.props().name).toEqual('type');
  });

  it('should have a submit input', () => {
    const wrapper = renderWithEnzymes();
    const button = wrapper.find("input[type='submit']");
    expect(button.exists()).toBe(true);
    expect(button.text()).toEqual('');
  });

  it('should render CreateAccount Component', () => {
    const wrapper = renderWithEnzymes();
    expect(toJson(wrapper)).toMatchSnapshot();
  });


  it('should submit a valid form', () => {
    const wrapper = renderWithEnzymes();
    const event = {
      preventDefault: jest.fn()
    };
    const typeInput = wrapper.find("select[name='type']");
    typeInput.simulate('change', {
      target: {
        name: 'type',
        value: 'savings'
      }
    });
   
    wrapper.find('form').simulate('submit', event);
    expect(wrapper.find('CreateAccount').state('type')).toEqual('savings');

  });
});