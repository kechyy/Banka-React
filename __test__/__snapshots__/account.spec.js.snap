// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateAccount should render CreateAccount Component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <withRouter(Connect(CreateAccount))
        accountAction={[MockFunction]}
        getAccountAction={[MockFunction]}
        isAuthenticated={true}
        userProfile={[MockFunction]}
      >
        <Connect(CreateAccount)
          accountAction={[MockFunction]}
          getAccountAction={[MockFunction]}
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "createHref": [Function],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          isAuthenticated={true}
          location={
            Object {
              "hash": "",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          userProfile={[MockFunction]}
        >
          <CreateAccount
            account={Array []}
            accountAction={[Function]}
            getAccountAction={[Function]}
            history={
              Object {
                "action": "POP",
                "block": [Function],
                "createHref": [Function],
                "go": [Function],
                "goBack": [Function],
                "goForward": [Function],
                "length": 1,
                "listen": [Function],
                "location": Object {
                  "hash": "",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
                "push": [Function],
                "replace": [Function],
              }
            }
            isAuthenticated={true}
            location={
              Object {
                "hash": "",
                "pathname": "/",
                "search": "",
                "state": undefined,
              }
            }
            match={
              Object {
                "isExact": true,
                "params": Object {},
                "path": "/",
                "url": "/",
              }
            }
            profileAction={[Function]}
            userProfile={Object {}}
          >
            <div>
              <Connect(MainHeader)>
                <MainHeader
                  logoutAction={[Function]}
                  profileAction={[Function]}
                >
                  <header
                    className="mainHeader"
                  >
                    <div
                      className="wrapper"
                    >
                      <div
                        className="container topHeader"
                      >
                        <div
                          className="row"
                          style={
                            Object {
                              "justifyContent": "flex-end",
                            }
                          }
                        >
                          <Link
                            className="logout btn btn-pink"
                            id="logout"
                            onClick={[Function]}
                            to="/signin"
                          >
                            <a
                              className="logout btn btn-pink"
                              href="/signin"
                              id="logout"
                              onClick={[Function]}
                            >
                              LOGOUT
                            </a>
                          </Link>
                          <button
                            className="topMenu btn btn-red"
                            onClick={[Function]}
                            style={
                              Object {
                                "padding": "10px",
                              }
                            }
                          >
                            Menu
                          </button>
                          <ul
                            className="dropDown"
                            style={
                              Object {
                                "display": "none",
                              }
                            }
                          >
                            <li
                              className="userProfiles"
                            >
                              <a
                                data-page="profile"
                                data-title="Profile Details"
                              >
                                Profile
                              </a>
                            </li>
                            <li
                              className="accountForms"
                            >
                              <span
                                className="fa fa-book"
                              />
                              <a
                                data-page="account"
                                data-title="Create Account"
                              >
                                Create Bank Account
                              </a>
                            </li>
                          </ul>
                        </div>
                      </div>
                    </div>
                  </header>
                </MainHeader>
              </Connect(MainHeader)>
              <main>
                <div
                  className="wrapper "
                >
                  <Connect(SideBar)>
                    <SideBar
                      account={Array []}
                      getTransactionAction={[Function]}
                      userProfile={Object {}}
                    >
                      <div
                        className="left mainSideBar"
                      >
                        <div
                          className="row"
                        >
                          <div
                            className="col-8-lg sideBar"
                          >
                            <aside>
                              <div>
                                <img
                                  alt=""
                                  className="logo"
                                  src="../images/logo2.png"
                                />
                              </div>
                              <div
                                className="sideBarTitle"
                              >
                                Client User
                              </div>
                              <ul
                                className="sideMenu"
                              >
                                <li
                                  className="userProfiles"
                                >
                                  -  
                                  <Link
                                    data-title="Profile Details"
                                    to="/user-profile"
                                  >
                                    <a
                                      data-title="Profile Details"
                                      href="/user-profile"
                                      onClick={[Function]}
                                    >
                                      Profile
                                    </a>
                                  </Link>
                                </li>
                                <li
                                  className="accountForms"
                                >
                                  <span
                                    className="fa fa-book"
                                  />
                                  <Link
                                    data-title="Create Account"
                                    to="/create-account"
                                  >
                                    <a
                                      data-title="Create Account"
                                      href="/create-account"
                                      onClick={[Function]}
                                    >
                                      Create Bank Account
                                    </a>
                                  </Link>
                                </li>
                                <li
                                  className="transact"
                                >
                                  -  
                                  <a
                                    onClick={[Function]}
                                  >
                                    Transaction History
                                  </a>
                                  <ul
                                    className="sideBarDropDown"
                                    style={
                                      Object {
                                        "display": "none",
                                      }
                                    }
                                  />
                                </li>
                              </ul>
                            </aside>
                          </div>
                        </div>
                      </div>
                    </SideBar>
                  </Connect(SideBar)>
                  <div
                    className="right"
                  >
                    <div
                      className="row"
                    >
                      <div
                        className="col-12-lg mainContent"
                      >
                        <h2
                          id="contentMainTitle"
                        >
                          User Account Page
                        </h2>
                        <header
                          className="mainTitle"
                        >
                          <div
                            className="row "
                          >
                            <div
                              className="col-12-md"
                            >
                              <h3
                                id="contentSubTitle"
                              >
                                Bank Account
                              </h3>
                            </div>
                          </div>
                        </header>
                        <aside
                          className="rightSideBar"
                        >
                          <div
                            className="row"
                            id="userMain"
                          >
                            <div
                              style={
                                Object {
                                  "width": "100%",
                                }
                              }
                            >
                              <div
                                className="col-8-md acctDetails"
                              >
                                <form
                                  onSubmit={[Function]}
                                >
                                  <div
                                    className="row createAcct"
                                  >
                                    <div
                                      className="col-6-md col-12-xs"
                                    >
                                      <label
                                        htmlFor="true"
                                      >
                                        First Name
                                      </label>
                                      <input
                                        className="input-block"
                                        disabled={true}
                                        type="text"
                                      />
                                    </div>
                                    <div
                                      className="col-6-md col-12-xs"
                                    >
                                      <label
                                        htmlFor="true"
                                      >
                                        Last Name
                                      </label>
                                      <input
                                        className="input-block"
                                        disabled={true}
                                        type="text"
                                      />
                                    </div>
                                    <div
                                      className="col-6-md col-12-xs"
                                    >
                                      <label
                                        htmlFor="true"
                                      >
                                        Email Address
                                      </label>
                                      <input
                                        className="input-block"
                                        disabled={true}
                                        type="email"
                                      />
                                    </div>
                                    <div
                                      className="col-6-md col-12-xs"
                                    >
                                      <label
                                        className="required"
                                        htmlFor="accountType"
                                      >
                                        Account type 
                                        <span>
                                          *
                                        </span>
                                      </label>
                                      <select
                                        className="input-block"
                                        id="type"
                                        name="type"
                                        onChange={[Function]}
                                        required={true}
                                      >
                                        <option
                                          value=""
                                        >
                                          --Select Account Type--
                                        </option>
                                        <option
                                          value="savings"
                                        >
                                          Savings
                                        </option>
                                        <option
                                          value="current"
                                        >
                                          Current
                                        </option>
                                      </select>
                                    </div>
                                    <div
                                      className="col-6-md col-12-xs"
                                    >
                                      <input
                                        className="btn btn-blue createAccountBtn"
                                        defaultValue="Create Account"
                                        id="createAccountBtn"
                                        type="submit"
                                      />
                                    </div>
                                  </div>
                                </form>
                                <div
                                  className="row "
                                >
                                  <div
                                    className="col-12-xs col-12-md table"
                                  >
                                    <table
                                      cellPadding={0}
                                      cellSpacing={0}
                                    >
                                      <thead>
                                        <tr>
                                          <th>
                                            S/N
                                          </th>
                                          <th>
                                            Account Number
                                          </th>
                                          <th>
                                            Account type
                                          </th>
                                          <th>
                                            Account Balance
                                          </th>
                                          <th>
                                            Account Status
                                          </th>
                                          <th>
                                            Created Date
                                          </th>
                                        </tr>
                                      </thead>
                                      <tbody />
                                    </table>
                                  </div>
                                </div>
                              </div>
                            </div>
                          </div>
                        </aside>
                      </div>
                    </div>
                  </div>
                </div>
              </main>
              <MainFooter>
                <footer
                  className="footer"
                >
                  <p>
                    Copyright © Kecyy Banka Application all 2019 rights reserved
                  </p>
                </footer>
              </MainFooter>
            </div>
          </CreateAccount>
        </Connect(CreateAccount)>
      </withRouter(Connect(CreateAccount))>
    </Router>
  </BrowserRouter>
</Provider>
`;
